capabilities/capabilities.go

```diff
@@ -0,0 +1,95 @@
+// Package capabilities provides an API for capability detection.
+package capabilities
+
+import (
+	"github.com/Masterminds/semver"
+)
+
+var (
+	// Autoscaling is the capability to scale tenant clusters automatically.
+	Autoscaling = CapabilityDefinition{
+		Name: "Autoscaling",
+		RequiredReleasePerProvider: []ReleaseProviderPair{
+			ReleaseProviderPair{
+				Provider:       "aws",
+				ReleaseVersion: semver.MustParse("6.3"),
+			},
+		},
+	}
+
+	// AvailabilityZones is the capability to spread the worker nodes of a tenant
+	// cluster over multiple availability zones.
+	AvailabilityZones = CapabilityDefinition{
+		Name: "AvailabilityZones",
+		RequiredReleasePerProvider: []ReleaseProviderPair{
+			ReleaseProviderPair{
+				Provider:       "aws",
+				ReleaseVersion: semver.MustParse("6.1"),
+			},
+		},
+	}
+
+	// NodePools is the capabilitiy to group tenant cluster workers logically.
+	NodePools = CapabilityDefinition{
+		Name: "NodePools",
+		RequiredReleasePerProvider: []ReleaseProviderPair{
+			ReleaseProviderPair{
+				Provider:       "aws",
+				ReleaseVersion: semver.MustParse("9"), // TODO: fix once the node pools release version is defined.
+			},
+		},
+	}
+
+	// AllCapabilityDefinitions contains all the capabilities
+	AllCapabilityDefinitions = []CapabilityDefinition{
+		Autoscaling,
+		AvailabilityZones,
+		NodePools,
+	}
+)
+
+// CapabilityDefinition is the type we use to describe the conditions that have to be met
+// so that we can assume a certain capability on the installation or API side.
+type CapabilityDefinition struct {
+	Name        string
+	Description string
+	// RequiredReleasePerProvider holds the combination(s) of provider and
+	// release version which have to be fulfilled so we assume a capability.
+	RequiredReleasePerProvider []ReleaseProviderPair
+}
+
+// ReleaseProviderPair is a combination of a providr ('aws', 'azure', 'kvm) and
+// a release version number.
+type ReleaseProviderPair struct {
+	Provider       string
+	ReleaseVersion *semver.Version
+}
+
+// GetCapabilities returns the capabilities available in the current context
+func GetCapabilities(provider string, releaseVersion *semver.Version) ([]CapabilityDefinition, error) {
+	cap := []CapabilityDefinition{}
+
+	// iterate all capabilities and find the ones that apply
+	for _, capability := range AllCapabilityDefinitions {
+		if HasCapability(provider, releaseVersion, capability) {
+			cap = append(cap, capability)
+		}
+	}
+
+	return cap, nil
+}
+
+// HasCapability returns true if the current context (provider, release) provides
+// the given capabililty.
+func HasCapability(provider string, releaseVersion *semver.Version, capability CapabilityDefinition) bool {
```

Comment:
```suggestion
func HasCapability(provider, releaseVersion string, capability CapabilityDefinition) bool {
```

I'd maybe accept the release version here as a string and perform the parsing inside. Two reasons:
* It doesn't tie this API to specific library (`semver`).
* By looking at code below it seems a bit ugly to do parsing before call to this.

I'm not fully sure which approach is better: Perform error logging & `os.Exit(1)` in here or return the error, but when release version is something that's quite deeply built into our product, I'd assume it's always an unrecoverable error when it cannot be parsed and therefore just terminating here would probably be correct.

Reply:
Thanks for the suggestion. I'll do the semver parsing inside of capabilities.

I wouldn't panic or exit in the `capabilities` package, but return the error instead. Treating that should be the responsibility of the command packages.

Reply:
Done